core/src/apps/cardano/sign_tx.py

```diff
@@ -100,526 +111,529 @@
+        await confirm_stake_pool_registration_final(ctx)
 
-    if is_mainnet_network_id != is_mainnet_protocol_magic:
-        raise wire.ProcessError("Invalid network id/protocol magic combination!")
 
+async def _process_inputs(
+    ctx: wire.Context,
+    msg: CardanoSignTxInit,
+    keychain: seed.Keychain,
+    builder: TxBuilder,
+) -> None:
+    """Read, validate and serialize the inputs."""
+    builder.start_inputs(msg.inputs_count)
+    for index in range(msg.inputs_count):
+        input: CardanoTxInput = await ctx.call(CardanoTxItemAck(), CardanoTxInput)
+        builder.add_input(input.prev_hash, input.prev_index)
 
-def _validate_stake_pool_registration_tx_structure(msg: CardanoSignTx) -> None:
-    # ensures that there is exactly one certificate, which is stake pool registration,
-    # and no withdrawals
-    if (
-        len(msg.certificates) != 1
-        or not _has_stake_pool_registration(msg)
-        or len(msg.withdrawals) != 0
-    ):
-        raise INVALID_STAKE_POOL_REGISTRATION_TX_STRUCTURE
+    builder.finish_inputs()
 
 
-def _validate_outputs(
+async def _process_outputs(
+    ctx: wire.Context,
+    msg: CardanoSignTxInit,
     keychain: seed.Keychain,
-    outputs: list[CardanoTxOutputType],
-    protocol_magic: int,
-    network_id: int,
-) -> None:
+    builder: TxBuilder,
+) -> int:
+    """Read, validate, confirm and serialize the outputs, return the total non-change output amount."""
+    builder.start_outputs(msg.outputs_count)
     total_amount = 0
-    for output in outputs:
-        total_amount += output.amount
-
-        if output.address_parameters and output.address is not None:
-            raise wire.ProcessError(
-                "Outputs can not contain both address and address_parameters fields!"
-            )
-
-        if output.address_parameters:
-            validate_address_parameters(output.address_parameters)
-        elif output.address is not None:
-            validate_output_address(output.address, protocol_magic, network_id)
+    total_amount_to_show = 0
+    for _ in range(msg.outputs_count):
+        output: CardanoTxOutput = await ctx.call(CardanoTxItemAck(), CardanoTxOutput)
+        _validate_output(msg, output)
+        if msg.signing_mode == CardanoTxSigningMode.ORDINARY_TRANSACTION:
+            amount_to_show = await _show_output(ctx, keychain, msg, output)
+            total_amount_to_show += amount_to_show or 0
+
+        output_address = _get_output_address(
+            keychain, msg.protocol_magic, msg.network_id, output
+        )
+        if output.asset_groups_count == 0:
+            builder.add_simple_output(output.amount, output_address)
         else:
-            raise wire.ProcessError(
-                "Each output must have an address field or address_parameters!"
-            )
+            builder.add_output_with_tokens(output.amount, output_address)
+            await _process_asset_groups(ctx, msg, builder, output.asset_groups_count)
```

Comment:
I've changed it so that the parameters are being passed in separately. Mostly `msg` was being passed in because of `singing_mode`, `protocol_magic` and `network_id`. I'm not 100% sure it's better to separate the parameters, but it's probably worth it for the decoupling of the functions from `msg` and making it clear which params they use.

Reply:
Conceptually, it's better not to pass around the tx init msg. But yes, a bit too many parameters now. Perhaps it would make sense to wrap some up into an object (e.g. SupplementaryTxData containing signing mode and those network params).